### PR TITLE
Automate DOI

### PR DESCRIPTION
**Pull request by shadrach-tayo**
## Checklist before Merge
- Publish new version of desci-models
- Ensure this [PR](https://github.com/desci-labs/nodes/pull/352) is merged

## Description of the Problem / Feature
- Attach DOI to manuscripts in draft mode if already assigned 
- Use Crossref client to manuscript for DOIs 
- Add field `doi` to desci-models pdf component payload

### CODE REVIEW

**Comment by shadrach-tayo** [comment]
@hubsmoke Can you publish a new version of desci-models? I need it for this PR to work

File: desci-server/src/controllers/doi/check.ts [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+  const componentIndex = latestManifest.components.findIndex(
+    (component) =>
+      component.type === ResearchObjectComponentType.PDF && (component.payload as CommonComponentPayload).path === path,
+  );
+
+  if (componentIndex === -1) throw new BadRequestError('Component to attach DOI not a valid pdf');
+
+  const component = latestManifest.components[componentIndex] as PdfComponent;
+
+  if (component.payload.doi) throw new ForbiddenError(`${component.subtype || component.type} already has a DOI`);
```
Comment (hubsmoke) [inline desci-server/src/controllers/doi/check.ts:67]: Make sure DOI is an array, there can be multiple DOIs for a pdf

File: desci-server/src/controllers/doi/check.ts [hunk 2]
```diff
@@ -25,3 +44,61 @@ export const getDoi = async (req: Request, res: Response, next: NextFunction) =>
...
+    component.name.replace(/\.pdf/g, '') ||
+    component.payload.title;
+
+  const works = await crossRefClient.listWorks({
+    queryTitle,
+    rows: 5,
+    select: [WorkSelectOptions.DOI, WorkSelectOptions.TITLE, WorkSelectOptions.AUTHOR],
+  });
+
+  const doi = works?.data?.message?.items.find((item) => item.title.some((t) => t === queryTitle));
```
Comment (m0ar) [inline desci-server/src/controllers/doi/check.ts:80]: fwiw, equality will probably lead to some false negatives. Not sure if it's worth refining at this point though.

File: desci-server/src/routes/v1/doi.ts [hunk 3]
```diff
@@ -1,11 +1,21 @@
...
+  ensureUser,
+  getDoi,
+  mintDoi,
+  validate,
+} from '../../internal.js';
 const router = Router();
 
 router.post('/check/:uuid', [ensureUser, ensureNodeAccess], asyncHander(checkMintability));
 router.post('/mint/:uuid', [ensureUser, ensureNodeAccess], asyncHander(mintDoi));
+router.post('/attach', [ensureUser, ensureNodeAccess, validate(attachDoiSchema)], asyncHander(attachDoi));
```
Comment (m0ar) [inline desci-server/src/routes/v1/doi.ts:18]: Not immediately clear what “attach” means in this case :thinking:

**Review by m0ar — APPROVED** [review]
:ok_hand: :shipit:  :ok_hand:

File: desci-server/src/routes/v1/doi.ts (hunk 3 above)
Comment (shadrach-tayo) [reply desci-server/src/routes/v1/doi.ts:18]: updates the manuscript in the manifest to add doi to the payload

File: desci-server/src/controllers/doi/check.ts (hunk 1 above)
Comment (shadrach-tayo) [reply desci-server/src/controllers/doi/check.ts:72]: I didn't catch this, I'll push the updates to a new branch

**Pull request merged**